Source/HTTPHeaders.swift

```diff
@@ -24,51 +24,215 @@
 
 import Foundation
 
-public typealias HTTPHeaders = [String: String]
+public struct HTTPHeaders {
+    private var headers = [HTTPHeader]()
+    
+    public init() { }
+    
+    public init(_ headers: [HTTPHeader]) {
+        self.headers = headers
+    }
+    
+    public init(_ dictionary: [String: String]) {
+        self.init()
+        
+        headers = dictionary.map { HTTPHeader(name: $0.key, value: $0.value) }
+    }
+    
+    public mutating func update(name: String, value: String) {
+        guard let index = headers.index(of: name) else {
+            headers.append(HTTPHeader(name: name, value: value))
+            return
+        }
+        
+        headers.replaceSubrange(index...index, with: [HTTPHeader(name: name, value: value)])
+    }
+    
+    public mutating func update(_ header: HTTPHeader) {
+        update(name: header.name, value: header.value)
+    }
+    
+    public mutating func remove(name: String) {
+        guard let index = headers.index(of: name) else { return }
+        
+        headers.remove(at: index)
+    }
+    
+    mutating func sort() {
+        return headers.sort()
+    }
+    
+    func sorted() -> HTTPHeaders {
+        return HTTPHeaders(headers.sorted())
+    }
+    
+    func value(for name: String) -> String? {
+        guard let index = headers.index(of: name) else { return nil }
+        
+        return headers[index].value
+    }
+    
+    public subscript(_ name: String) -> String? {
+        get { return value(for: name) }
+        set {
+            if let value = newValue {
+                update(name: name, value: value)
+            } else {
+                remove(name: name)
+            }
+        }
+    }
+}
+
+extension HTTPHeaders: ExpressibleByDictionaryLiteral {
+    public init(dictionaryLiteral elements: (String, String)...) {
+        self.init()
+        
+        elements.forEach { update(name: $0.0, value: $0.1) }
+    }
+    
+    public var dictionary: [String: String] {
+        let namesAndValues = headers.map { ($0.name, $0.value) }
+        
+        return Dictionary(namesAndValues, uniquingKeysWith: { (_, last) in last })
+    }
+}
+
+extension HTTPHeaders: ExpressibleByArrayLiteral {
+    public init(arrayLiteral elements: HTTPHeader...) {
+        self.init()
```

Comment:
If deduplication is to be performed in `init(_ headers: [HTTPHeader])`, this can be changed to: `self.init(elements)`

Reply:
Done.